src/libraries/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs

```diff
@@ -31,168 +31,287 @@ protected static void ThrowMultipleContinuations() =>
+#if NET9_0_OR_GREATER
+                    false;
+#else
+                    0;
+#endif
             }
 
-            if (!IsCompleted)
+            Debug.Assert(_cancellationRegistration == default);
+            return true;
+        }
+
+        /// <summary>Completes the operation with a failed state and the specified error.</summary>
+        /// <param name="exception">The error.</param>
+        /// <returns>true if the operation could be successfully transitioned to a completed state; false if it was already completed.</returns>
+        public bool TrySetException(Exception exception)
+        {
+            if (TryReserveCompletionIfCancelable())
             {
-                ThrowIncompleteOperationException();
+                AssertCompletionReserved();
+                _error = ExceptionDispatchInfo.Capture(exception);
+                SignalCompletion();
+                return true;
             }
 
-            ExceptionDispatchInfo? error = _error;
-            TResult? result = _result;
-            _currentId++;
+            return false;
+        }
 
-            if (_pooled)
+        /// <summary>Completes the operation with a failed state and a cancellation error.</summary>
+        /// <param name="cancellationToken">The cancellation token that caused the cancellation.</param>
+        /// <returns>true if the operation could be successfully transitioned to a completed state; false if it was already completed.</returns>
+        public bool TrySetCanceled(CancellationToken cancellationToken = default)
+        {
+            if (TryReserveCompletionIfCancelable())
             {
-                Volatile.Write(ref _continuation, s_availableSentinel); // only after fetching all needed data
+                AssertCompletionReserved();
+                _error = ExceptionDispatchInfo.Capture(new OperationCanceledException(cancellationToken));
+                SignalCompletion();
+                return true;
             }
 
-            error?.Throw();
-            return result!;
+            return false;
         }
 
-        /// <summary>Gets the result of the operation.</summary>
-        /// <param name="token">The token that must match <see cref="_currentId"/>.</param>
-        void IValueTaskSource.GetResult(short token)
+        /// <summary>Attempts to reserve this instance for completion.</summary>
+        /// <remarks>
+        /// This will always return true for non-cancelable objects, as they only ever have a single owner
+        /// responsible for completion.  For cancelable operations, this will attempt to atomically transition
+        /// to a reserved completion state.
+        /// </remarks>
+        public bool TryReserveCompletionIfCancelable() =>
+            !CancellationToken.CanBeCanceled ||
+            Interlocked.CompareExchange(ref _completionReserved,
+#if NET9_0_OR_GREATER
+                true, false) == false;
+#else
+                1, 0) == 0;
+#endif
```

Comment:
Exchange should be enough:

```suggestion
            Interlocked.Exchange(ref _completionReserved,
#if NET9_0_OR_GREATER
                true) == false;
#else
                1) == 0;
#endif
```